core/src/main/java/com/yahoo/oak/NovaMMHeader.java

```diff
@@ -0,0 +1,152 @@
+    *
+    * The length (Mostly in integer size) of the data is also held just following the header (since the block and
+    * offset are 42 bits at most the length can be 42 bits which we allow here!)
+    * The length is set once upon header allocation and later can only be read. */
+
+    private static final int VERSION_SIZE = 22;  //block offset get 42 bits on heap-(including one bit for deleted)
+   
+    private static final int OFFHEAP_VERSION_MASK = 0x3F_FF_FF;
+
+
+    int getOffHeapVersion(long headerAddress) {
+        return (int) (UnsafeUtils.UNSAFE.getLong(headerAddress) & OFFHEAP_VERSION_MASK);
+    }
+
+    int getDataLength(long headerAddress) {
+        return (int) (UnsafeUtils.UNSAFE.getLong(headerAddress) >>> VERSION_SIZE); //TODO length can be long??
+    }
+    
+    long getOffHeapMetaData(long headerAddress) {
+        return UnsafeUtils.UNSAFE.getLong(headerAddress);
+    }
+
+    private void setDataLength(long headerAddress, int length) {
+        throw new IllegalAccessError();
+    }
+
+    void initHeader(long headerAddress, int dataLength, int version) {
+        long sliceHeader    = (long) (dataLength) << VERSION_SIZE;
+        int  newVer         = version & 0x3FFFFF; //VERSION_SIZE;
+        long header         = sliceHeader | newVer ;
+        UnsafeUtils.UNSAFE.putLong(headerAddress, header);
+    }
+
+    public  void setTap(AtomicLongArray tap, long ref, int idx) {
+        int i = idx % ThreadIndexCalculator.MAX_THREADS;
+        tap.set(NovaMemoryManager.CACHE_PADDING * (i + 1) + NovaMemoryManager.IDENTRY, idx);
+        tap.set(NovaMemoryManager.CACHE_PADDING * (i + 1) + NovaMemoryManager.REFENTRY, ref);
+    }
+        
+    public  void unsetTap(AtomicLongArray tap, int idx) {
+        int i = idx % ThreadIndexCalculator.MAX_THREADS;
+        tap.set(NovaMemoryManager.CACHE_PADDING * (i + 1) + NovaMemoryManager.IDENTRY, -1);
+    }
+
+
+    /*---------------- Locking Implementation ----------------*/
+
+    private static boolean cas(long headerAddress, int expectedLock, int newLock, int version) {
+        // Since the writing is done directly to the memory, the endianness of the memory is important here.
+        // Therefore, we make sure that the values are read and written correctly.
+        long expected = UnsafeUtils.intsToLong(version, expectedLock);
+        long value = UnsafeUtils.intsToLong(version, newLock);
+        return UnsafeUtils.UNSAFE.compareAndSwapLong(null, headerAddress, expected, value);
+    }
+
+    
+    private static boolean cas(long headerAddress, long offHeapMeta) {
+        // Since the writing is done directly to the memory, the endianness of the memory is important here.
+        // Therefore, we make sure that the values are read and written correctly.
+        return UnsafeUtils.UNSAFE.compareAndSwapLong(null, headerAddress, offHeapMeta, offHeapMeta | 1);
+    }
+    
+    
+    ValueUtils.ValueResult lockRead(final int onHeapVersion, long headerAddress) {
```

Comment:
It is not really a lock/unlock so, as I previously mentioned, confusing... Please think about new (additional) API for start/stop in Slide.

Also checking the bit via `%2` is not clear. Either comment very well, ore use `UnionCodec` for three values